File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/ViewModel/ViewModelsInjectOnlyUseCasesAndMappers.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class ViewModelsInjectOnlyUseCasesAndMappers: QuickSpec {
    override class func spec() {
        Given("A ViewModel initializer parameter") {
            let viewModels = WoofHarmonize.viewModels

            Then("Every dependency is a UseCase, a Mapper, the NavigatorImplementing, or a @DIArgument value") {
                viewModels.assertTrue(message: message) { viewModel in
                    let arguments = Set(viewModel.variables.filter { $0.hasAttribute(named: "@DIArgument") }.map(\.name))
                    return viewModel.initializers.flatMap(\.parameters).allSatisfy { parameter in
                        let type = parameter.typeAnnotation?.name ?? ""
                        return type.hasSuffix("UseCase")
                            || type.hasSuffix("Mapper")
                            || type == "NavigatorImplementing"
                            || arguments.contains(parameter.name)
                    }
                }
            }
        }
    }

    private static let message = LintRuleMessage(
        rule: "ViewModels may only inject UseCases, Mappers, NavigatorImplementing (for state-driven navigation), and @DIArgument values.",
        why: """
            ViewModels orchestrate use cases and shape their output for the UI. Any other dependency
            (repositories, data sources, framework classes) belongs to a lower layer.
            """,
        howToFix: "Move the work into a UseCase or a Mapper and inject that.",
        badExample: "final class ProfileWoofViewModel { private let woofsRepository: WoofsRepository }",
        goodExample: "final class ProfileWoofViewModel { @DIArgument private let dog: Dog; private let sendWoofUseCase: SendWoofUseCase }"
    )
}

File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/viewmodel/ViewModelsInjectOnlyUseCasesAndMappers.kt
package com.perrystreet.woof.konsist.viewmodel

import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class ViewModelsInjectOnlyUseCasesAndMappers : BehaviorSpec() {
    init {
        Given("A ViewModel constructor") {
            val parameters = KonsistUtils.viewModels.flatMap { it.primaryConstructor?.parameters.orEmpty() }

            Then("Every dependency is a UseCase, a Mapper, the INavigator, or an injected parameter") {
                parameters.assertTrue(message = Message) { parameter ->
                    parameter.type.name.endsWith("UseCase") ||
                        parameter.type.name.endsWith("Mapper") ||
                        parameter.type.name == "INavigator" ||
                        parameter.hasAnnotationWithName("InjectedParam")
                }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "ViewModels may only inject UseCases, Mappers, INavigator (for state-driven navigation), and @InjectedParam values.",
            why = """
                ViewModels orchestrate use cases and shape their output for the UI. Any other dependency
                (repositories, data sources, framework classes) belongs to a lower layer.
            """.trimIndent(),
            howToFix = "Move the work into a UseCase or a Mapper and inject that.",
            badExample = "class ProfileWoofViewModel(private val woofsRepository: WoofsRepository)",
            goodExample = "class ProfileWoofViewModel(@InjectedParam private val dog: Dog, private val sendWoofUseCase: SendWoofUseCase)",
        )
    }
}
